### DISCUSSION TITLE
v1.29.0: errors using `render()` after React v18 update

### DISCUSSION THREAD

**Discussion by Rodrigoplp-work** [eclipse-theia/theia#11620] [Q&A]
After updating the dependency of `@theia/core` to 1.29.0, this error is being thrown for classes that extend Theia's react-widget:

"Property 'render' in (...) is not assignable to the same property in base type 'ReactWidget'."

For example, the following class:

```tsx
import { ReactWidget } from '@theia/core/lib/browser';
import * as React from 'react';

export class TraceExplorerViewsWidget extends ReactWidget {
  ...
  render(): React.ReactNode {
    return
      <div>
        <ReactAvailableViewsWidget
          id={this.id}
          title={this.title.label}
          tspClientProvider={this.tspClientProvider}
        >
        </ReactAvailableViewsWidget>
      </div>
  }
}
```

Throws this error on project build:

```sh
Property 'render' in type 'TraceExplorerViewsWidget' is not assignable to the same property in base type 'ReactWidget'.
  Type '() => import("[...]/node_modules/@types/react/index").ReactNode' is not assignable to type '() => React.ReactNode'.
    Type 'import("[...]/node_modules/@types/react/index").ReactNode' is not assignable to type 'React.ReactNode'.
      Type '{}' is not assignable to type 'ReactNode'.
        Type '{}' is missing the following properties from type 'ReactPortal': key, children, type, props
```

In this case, how could `render(): React.ReactNode` be changed in order to use the new `createRoot` API?

**Comment by paul-marechal** [comment]
Can you run `yarn why @types/react`?

It looks like conflicting typings coming out of duplicated `@types/react` packages.

**Reply by paul-marechal** [reply]
Looking at some code like https://github.com/eclipse-cdt-cloud/theia-trace-extension/blob/193a827083978af67de7785b3df56b150b762ece/theia-extensions/viewer-prototype/src/browser/trace-explorer/trace-explorer-sub-widgets/theia-trace-explorer-opened-traces-widget.tsx#L3

I think you should replace those direct references to the `react` package with `@theia/core/shared/react`. This way you will be sure to use the package and types used by the framework.

**Reply by Rodrigoplp-work** [reply]
Yes, we reference 3 different versions of React. I am hoping not to have to upgrade the older ones now, because we use `render(): React.ReactNode {}` everywhere.
The case shown above, though, uses the version 18 from Theia. If we find a way to make `render()` work there, then we can plan to extend the update for the rest of the project.
I was also hoping to be able to make it work only with changes on our side, and not have to change Theia's ReactWidget line 50 from
```tsx
protected abstract render(): React.ReactNode;
```
to something like
```tsx
protected abstract render(node: React.ReactNode): void {
  this.nodeRoot.render(node);
}
```

**Reply by paul-marechal** [reply]
fyi the following error you reported comes from the return type in `render(): React.ReactNode` which is conflicting between the react typings expected by the Theia framework and the typings you happen to use in your extension:

```
Property 'render' in type 'TraceExplorerViewsWidget' is not assignable to the same property in base type 'ReactWidget'.
  Type '() => import("[...]/node_modules/@types/react/index").ReactNode' is not assignable to type '() => React.ReactNode'.
    Type 'import("[...]/node_modules/@types/react/index").ReactNode' is not assignable to type 'React.ReactNode'.
      Type '{}' is not assignable to type 'ReactNode'.
        Type '{}' is missing the following properties from type 'ReactPortal': key, children, type, props
```

> I was also hoping to be able to make it work only with changes on our side, and not have to change Theia's ReactWidget [...]

I don't think any changes are required in Theia.

My advice for your Theia extension is to not depend on react directly, and instead leverage only the `@theia/core` package and import react APIs using `import ... from '@theia/core/shared/react`. This way it will precisely pull what's expected by the framework.
